ggshield/config.py

```diff
@@ -23,139 +24,198 @@ def replace_in_keys(data: Dict, old_char: str, new_char: str) -> None:
-        ]
-
-        for attr in self.attributes:
-            setattr(self, attr.name, attr.default)
-        self.load_configs(self.CONFIG_GLOBAL)
-        self.load_configs(self.CONFIG_LOCAL)
+def load_yaml(path: str, raise_exc: bool = False) -> Optional[Dict[str, Any]]:
+    if not os.path.isfile(path):
+        return None
 
-    def __getattr__(self, name: str) -> Any:
-        # Required for dynamic types on mypy
-        return object.__getattribute__(self, name)
-
-    def get_attributes_keys(self) -> List:
-        return list(
-            list(zip(*self.attributes))[0]
-        )  # get list of first elements in tuple
-
-    def update_config(self, **kwargs: Any) -> None:
-        for key, item in kwargs.items():
-            if key in self.get_attributes_keys():
-                if isinstance(getattr(self, key), list):
-                    getattr(self, key).extend(item)
-                elif isinstance(getattr(self, key), set):
-                    getattr(self, key).update(item)
-                else:
-                    setattr(self, key, item)
+    with open(path, "r") as f:
+        try:
+            data = yaml.safe_load(f) or {}
+        except Exception as e:
+            message = f"Parsing error while reading {path}:\n{str(e)}"
+            if raise_exc:
+                raise click.ClickException(message) from e
             else:
-                click.echo("Unrecognized key in config: {}".format(key))
+                click.echo(message)
+                return None
+        else:
+            return data
 
-    def load_config(self, filename: str) -> bool:
-        if not os.path.isfile(filename):
-            return False
 
-        with open(filename, "r") as f:
-            try:
-                _config = yaml.safe_load(f) or {}
-                replace_in_keys(_config, "-", "_")
-                self.update_config(**_config)
-            except Exception as e:
-                raise click.ClickException(
-                    f"Parsing error while reading {filename}:\n{str(e)}"
-                )
+def get_global_path(filename: str) -> str:
+    return os.path.join(os.path.expanduser("~"), filename)
 
-        return True
 
-    def load_configs(self, filenames: List[str]) -> None:
-        """Loads config files until one succeeds."""
-        for filename in filenames:
-            try:
-                if self.load_config(filename):
-                    return
-            except Exception as exc:
-                click.echo(str(exc))
-
-    def to_dict(self) -> Dict[str, Any]:
-        _config = {key: getattr(self, key) for key in self.get_attributes_keys()}
-        # Convert all sets into more human readable lists
-        for key in self.get_attributes_keys():
-            value = _config[key]
-            if type(value) is set:
-                _config[key] = list(value)
-        replace_in_keys(_config, "_", "-")
-        return _config
-
-    def save(self) -> bool:
+class YamlFileConfig:
```

Comment:
why not using abstract class?

Reply:
why should we use it here? it's useful to expose abstraction from a module, but I don't really find it useful inside one module.